app/src/main/java/com/lambdasoup/quickfit/alarm/AlarmService.kt

```diff
@@ -0,0 +1,177 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.lambdasoup.quickfit.alarm
+
+import android.app.*
+import android.content.Context
+import android.content.Intent
+import android.media.AudioAttributes
+import android.net.Uri
+import android.os.Build
+import android.os.IBinder
+import android.os.PowerManager
+import android.provider.Settings
+import androidx.core.app.NotificationCompat
+import androidx.preference.PreferenceManager
+import com.lambdasoup.quickfit.Constants
+import com.lambdasoup.quickfit.Constants.PENDING_INTENT_WORKOUT_LIST
+import com.lambdasoup.quickfit.R
+import com.lambdasoup.quickfit.persist.QuickFitContentProvider
+import com.lambdasoup.quickfit.ui.WorkoutListActivity
+import com.lambdasoup.quickfit.util.WakefulIntents
+import timber.log.Timber
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+/**
+ * ForegroundService, handling alarm-related background I/O work. Could also have been solved as a [androidx.core.app.JobIntentService],
+ * or as a collection of jobs for [androidx.work.WorkManager], but (a) reading the documentation says that JobScheduler is for _deferrable_
+ * I/O work, which this is not - and also no promises are given as to the timeliness of the execution of JobScheduler jobs that are free
+ * of restrictions. Although it seems to be fine in practice (on non-background crippling devices). Also (b) - just for trying it out.
+ * Currently, this project serves as an exhibition of different ways of performing background work.
+ */
+class AlarmService : Service() {
+    private val alarms by lazy { Alarms(this.applicationContext) }
+    private val runWakeLock by lazy {
+        (getSystemService(Context.POWER_SERVICE) as PowerManager)
+                .newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, AlarmService::class.java.canonicalName)
+    }
+
+    private lateinit var executor: ExecutorService
+
+    override fun onCreate() {
+        super.onCreate()
+        executor = Executors.newFixedThreadPool(2)
+    }
+
+    override fun onDestroy() {
+        executor.shutdownNow()
+        super.onDestroy()
+    }
+
+    override fun onBind(intent: Intent?): IBinder? = null
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        @Suppress("ReplaceGuardClauseWithFunctionCall") // IDE fails then, demands wrong type. New type inference at fault?
+        if (intent == null) throw IllegalArgumentException("Should never receive null intents because of START_REDELIVER_INTENT")
+
+        startForeground(Constants.NOTIFICATION_ALARM_BG_IO_WORK, buildForegroundNotification())
```

Comment:
maybe just call it `startForeground()` and do everything in there.

Reply:
not sure what you mean - but regardless, I'd start abstracting this kind of service only when I have at least another one. Then, I definitely see the point in abstracting everything out up to the inner `when(intent.action)` or whatever intent processing block.

Reply:
I am just talking about the stuff with the builder method there. Just put it all away into a method which makes more sense.